### ISSUE TITLE
NPE when Signal syncs contacts

### ISSUE THREAD

**Issue by bluebrother** [WhisperSystems/Signal-Android#5639]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

---
### Bug description

Signal force closes with a NPE when trying to sync contacts.
### Steps to reproduce

There are two ways to reproduce the issue:
- Go to Settings / Accounts / Signal
- Choose "Sync now" from the menu
- Shortly after a "Signal has been closed" from Android appears
- Open Signal, to to Settings and retrieve log.

or
- Open Signal, start a new conversation
- When selecting a contact to start conversation to, choose "Sync now" from the menu
- Shortly after a "Signal has been closed" from Android appears

**Actual result:** 
Looking at the logs shows a NPE when processing the contacts. Unfortunately the crash doesn't indicate which contact is causing the issue. I've also seen it after completely emptying the address book, so it looks like it's related to Signals contact mapping.

As a side note, after I removed and restored my address book I have several contacts in Signal that only show numbers, but not names. When checking the contact the Signal entry is missing. I can manually restore that to make the name show up again. My guess is that those names don't have the Signal entry (restored) because the sync crashed.

```
--------- beginning of crash
08-27 20:08:38.880 11357 11514 E AndroidRuntime: FATAL EXCEPTION: AsyncTask #3
08-27 20:08:38.880 11357 11514 E AndroidRuntime: Process: org.thoughtcrime.securesms, PID: 11357
08-27 20:08:38.880 11357 11514 E AndroidRuntime: java.lang.RuntimeException: An error occurred while executing doInBackground()
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at android.os.AsyncTask$3.done(AsyncTask.java:309)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.util.concurrent.FutureTask.finishCompletion(FutureTask.java:354)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.util.concurrent.FutureTask.setException(FutureTask.java:223)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.util.concurrent.FutureTask.run(FutureTask.java:242)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at android.os.AsyncTask$SerialExecutor$1.run(AsyncTask.java:234)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.lang.Thread.run(Thread.java:818)
08-27 20:08:38.880 11357 11514 E AndroidRuntime: Caused by: java.lang.NullPointerException: Attempt to invoke virtual method 'boolean java.lang.String.contains(java.lang.CharSequence)' on a null object reference
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.whispersystems.signalservice.api.util.PhoneNumberFormatter.formatNumber(PhoneNumberFormatter.java:79)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.contacts.ContactsDatabase.getSignalRawContacts(ContactsDatabase.java:330)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.contacts.ContactsDatabase.setRegisteredUsers(ContactsDatabase.java:88)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.util.DirectoryHelper.updateContactsDatabase(DirectoryHelper.java:203)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.util.DirectoryHelper.refreshDirectory(DirectoryHelper.java:101)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.util.DirectoryHelper.refreshDirectory(DirectoryHelper.java:72)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.ContactSelectionActivity$RefreshDirectoryTask.doInBackground(ContactSelectionActivity.java:141)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at org.thoughtcrime.securesms.ContactSelectionActivity$RefreshDirectoryTask.doInBackground(ContactSelectionActivity.java:126)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at android.os.AsyncTask$2.call(AsyncTask.java:295)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  at java.util.concurrent.FutureTask.run(FutureTask.java:237)
08-27 20:08:38.880 11357 11514 E AndroidRuntime:  ... 4 more
```

**Expected result:** Signal syncs successful and doesn't crash.
### Device info

Device  : motorola MotoG3 (osprey_reteu_2gb)
Android : 6.0 (36, MPI24.65-33)
Memory  : 15M (15.00% free, 96M max)
Memclass: 96
OS Host : ilclbld73
App     : Signal 3.17.0
### Link to debug log

https://gist.github.com/anonymous/c0513fed3187e1f3511b9aeba0bcd799

(Note: this shows different timestamps since I grabbed it later when reproducing the issue again. There is no change in the actual issue.)

**Comment by bluebrother** [comment]
I've managed to resolve (but not fix) this issue:
- Unregister from Signal
- Remove Signal account in Android account settings (I've hesitated to do that before, since it wasn't clear to me if that also removes my existing conversations. I figured from another issue that removing the account doesn't remove my existing conversations)
- Register to Signal again.

This has the drawback that my key changes but I can live with that (it's still somewhat unwanted).
Also, I first tried to unregister and re-register afterwards, this didn't work (same crash on the last step, same data in the log). Removing the Signal account from Android was necessary.

I'm keeping this open since I still consider it a bug -- Signal shouldn't crash if something is wrong with the account. It would be better to at least show some message that the  account is broken and re-registering is needed. I let it to Signal developers to decide if they thing this shouldn't be treated as a bug.
